File: rs2-core/src/pipeline/spec.rs
//! Typed pipeline spec (PRD §8.1). The stored format is structured and
//! schema-validatable; the terse string DSL is accepted as sugar and
//! converted on the way in ([`super::dsl`]).

use serde::{Deserialize, Serialize};

use crate::error::RsError;
use crate::retry::{EffectClass, RetryPolicy};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Mode {
    #[default]
    Serial,
    Parallel,
    /// First step whose condition passes executes; its result ends the pipeline.
    Conditional,
    /// Fire-and-forget branch; the original message continues immediately.
    Tee,
    /// Branch executes to completion, result discarded; original continues.
    TeeWait,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Action {
    /// Abort the pipeline; the failing message is the output.
    Stop,
    /// Continue to the next step.
    Next,
    /// The message exits the pipeline here (skips remaining steps).
    End,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Joiner {
    /// Named messages → one JSON object keyed by message name.
    JsonObject,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Splitter {
    /// JSON array/object body → one message per element.
    JsonSplit,
}

/// An internal or external call made by a pipeline step.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallSpec {
    pub method: String,
    /// Target URL; `${...}` interpolates variables and input-URL parts.
    pub url: String,
    /// Declared effect class; defaults from the method (PRD §7.1).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub effect: Option<EffectClass>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub headers: Option<std::collections::HashMap<String, String>>,
}

impl CallSpec {
    pub fn effect_class(&self) -> EffectClass {
        self.effect.unwrap_or_else(|| {
            http::Method::from_bytes(self.method.as_bytes())
                .map(|m| EffectClass::default_for_method(&m))
                .unwrap_or(EffectClass::Unsafe)
        })
    }
}

/// One pipeline step. Exactly one of `call` / `transform` / `pipeline` /
/// `split` must be present ([`PipelineSpec::validate`]).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct Step {
    /// Condition (PRD §8.2 expression language); step skipped when false.
    #[serde(rename = "if", skip_serializing_if = "Option::is_none")]
    pub condition: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub call: Option<CallSpec>,
    /// JSONata transform template applied to the JSON body.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transform: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pipeline: Option<Box<PipelineSpec>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub split: Option<Splitter>,
    /// Convert a failure into `{ _errorStatus, _errorMessage }` and continue.
    #[serde(rename = "try", skip_serializing_if = "std::ops::Not::not")]
    pub try_mode: bool,
    /// Run the step's call as **trusted internal composition**: the caller
    /// principal is dropped, so the call is authorized as an unattributed
    /// internal hop and can reach a mount the caller's roles could not (the
    /// gateway pattern — lock a service to e.g. `writeRoles: "A"` and front it
    /// with a caller-accessible pipeline that elevates). The target must admit
    /// internal access (a role-spec mount admits no-principal internal calls).
    /// Applies to `call` steps; default off. The authoring surface is
    /// `manageRoles`-gated, so only privileged authors can create elevations.
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    pub elevate: bool,
    /// `as: "$name"` captures the step result into a variable; the in-flight
    /// message keeps its prior body.
    #[serde(rename = "as", skip_serializing_if = "Option::is_none")]
    pub capture: Option<String>,
    /// Message name, keying `jsonObject` joins.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Per-call retry policy override (PRD §7.3 resolution).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub retry: Option<RetryPolicy>,
}

impl Step {
    pub fn call(method: &str, url: &str) -> Step {
        Step {
            call: Some(CallSpec {
                method: method.to_string(),
                url: url.to_string(),
                effect: None,
                headers: None,
            }),
            ..Default::default()
        }
    }

    /// The effect class this step contributes to segment analysis.
    pub fn effect_class(&self) -> Option<EffectClass> {
        self.call.as_ref().map(|c| c.effect_class())
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct PipelineSpec {
    pub mode: Mode,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub on_fail: Option<Action>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub on_succeed: Option<Action>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub concurrency: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub join: Option<Joiner>,
    pub steps: Vec<Step>,
}

impl PipelineSpec {
    /// Effective on-fail action: configured, else `stop` for parallel,
    /// `next` otherwise (Restspace semantics retained).
    pub fn fail_action(&self) -> Action {
        self.on_fail.unwrap_or(match self.mode {
            Mode::Parallel => Action::Stop,
            _ => Action::Next,
        })
    }

    pub fn succeed_action(&self) -> Action {
        self.on_succeed.unwrap_or(match self.mode {
            Mode::Conditional => Action::End,
            _ => Action::Next,
        })
    }

    /// Config-time validation: step shape, condition grammar, mode rules,
    /// nested pipelines. Returns all problems, not just the first.
    pub fn validate(&self) -> Vec<String> {
        let mut errors = Vec::new();
        self.validate_into("", &mut errors);
        errors
    }

    fn validate_into(&self, at: &str, errors: &mut Vec<String>) {
        if self.steps.is_empty() {
            errors.push(format!("{at}: pipeline has no steps"));
        }
        if self.mode == Mode::Parallel && self.on_fail.is_some_and(|a| a != Action::Stop) {
            errors.push(format!("{at}: parallel mode must stop on fail"));
        }
        if self.on_succeed == Some(Action::Stop) {
            errors.push(format!("{at}: cannot stop on success or the pipeline always fails"));
        }
        for (i, step) in self.steps.iter().enumerate() {
            let here = format!("{at}/steps[{i}]");
            let actions = [
                step.call.is_some(),
                step.transform.is_some(),
                step.pipeline.is_some(),
                step.split.is_some(),
            ]
            .iter()
            .filter(|p| **p)
            .count();
            if actions != 1 {
                errors.push(format!(
                    "{here}: a step must have exactly one of call/transform/pipeline/split"
                ));
            }
            if let Some(cond) = &step.condition {
                if let Err(e) = super::condition::Condition::parse(cond) {
                    errors.push(format!("{here}: invalid condition '{cond}': {e}"));
                }
            }
            if let Some(capture) = &step.capture {
                if !capture.starts_with('$') {
                    errors.push(format!("{here}: 'as' must name a variable starting with '$'"));
                }
            }
            if let Some(call) = &step.call {
                if http::Method::from_bytes(call.method.as_bytes()).is_err() {
                    errors.push(format!("{here}: invalid method '{}'", call.method));
                }
            }
            if step.elevate && step.call.is_none() {
                errors.push(format!("{here}: 'elevate' applies only to call steps"));
            }
            if let Some(sub) = &step.pipeline {
                sub.validate_into(&here, errors);
            }
        }
    }

    /// Parse from JSON: the typed form directly, or an array (the Restspace
    /// string DSL, PRD §8.1) converted via [`super::dsl::convert`].
    pub fn from_value(value: &serde_json::Value) -> Result<PipelineSpec, RsError> {
        let spec = if value.is_array() {
            super::dsl::convert(value)?
        } else {
            serde_json::from_value(value.clone())
                .map_err(|e| RsError::bad_request(format!("invalid pipeline spec: {e}")))?
        };
        let errors = spec.validate();
        if !errors.is_empty() {
            return Err(RsError::validation_failed(
                "pipeline spec failed validation",
                serde_json::json!(errors),
            ));
        }
        Ok(spec)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_the_prd_example() {
        let spec: PipelineSpec = serde_json::from_value(serde_json::json!({
            "mode": "serial",
            "onFail": "stop",
            "concurrency": 12,
            "steps": [
                { "call": { "method": "GET", "url": "/data/orders/${id}" }, "as": "$order" },
                { "if": "$order.status == 'open'",
                  "call": { "method": "POST", "url": "/payments/charge", "effect": "keyed" },
                  "try": true },
                { "transform": { "total": "$sum($order.lines.price)", "charged": "$_ok" } },
                { "pipeline": {
                    "mode": "parallel",
                    "steps": [
                        { "call": { "method": "GET", "url": "/data/customers/${order.customerId}" }, "as": "$customer" },
                        { "call": { "method": "GET", "url": "/stock/check" }, "as": "$stock" }
                    ],
                    "join": "jsonObject" } }
            ]
        }))
        .unwrap();
        assert_eq!(spec.steps.len(), 4);
        assert_eq!(spec.on_fail, Some(Action::Stop));
        assert_eq!(spec.steps[1].call.as_ref().unwrap().effect_class(), EffectClass::Keyed);
        assert!(spec.steps[1].try_mode);
        assert!(spec.validate().is_empty());
        // Method-inferred effects.
        assert_eq!(spec.steps[0].call.as_ref().unwrap().effect_class(), EffectClass::Pure);
    }

    #[test]
    fn validation_catches_shape_errors() {
        let spec: PipelineSpec = serde_json::from_value(serde_json::json!({
            "steps": [
                { "as": "$x" },
                { "if": "status ==", "call": { "method": "GET", "url": "/x" } },
                { "call": { "method": "NOT A METHOD", "url": "/x" } }
            ]
        }))
        .unwrap();
        let errors = spec.validate();
        assert_eq!(errors.len(), 3, "{errors:?}");
    }
}

File: rs2-worker/src/pipeline/spec.ts
// Typed pipeline spec (PRD §8.1). The stored format is structured and
// schema-validatable; the terse string DSL is accepted as sugar and
// converted on the way in (`./dsl`). Port of `rs2-core/src/pipeline/spec.rs`
// — `parse` reproduces serde's leniency (unknown keys ignored, wrong types
// rejected) and `toJson` its field order and `skip_serializing_if` rules,
// so a canonical spec stored by either host reads back identically.

import { RsError } from "../runtime/error";
import type { Json, JsonObject } from "../runtime/error";
import { validate as validatePattern } from "../runtime/path-pattern";
import { effectForMethod, retryFromConfig } from "../runtime/retry";
import type { EffectClass, RetryPolicy } from "../runtime/retry";
import { parseCondition } from "./condition";
import { validateExpr } from "./transform";

export type Mode = "serial" | "parallel" | "conditional" | "tee" | "teeWait";
export type Action = "stop" | "next" | "end";
export type Joiner = "jsonObject";
export type Splitter = "jsonSplit";

const MODES: Mode[] = ["serial", "parallel", "conditional", "tee", "teeWait"];
const ACTIONS: Action[] = ["stop", "next", "end"];
const EFFECTS: EffectClass[] = ["pure", "idempotent", "keyed", "unsafe"];

/// An internal or external call made by a pipeline step.
export interface CallSpec {
  method: string;
  /// Target URL; `${...}` interpolates variables and input-URL parts.
  url: string;
  effect: EffectClass | undefined;
  /// Extra request headers; values `${...}`-interpolate like the URL.
  headers: Record<string, string> | undefined;
}

/// Declared effect class; defaults from the method (PRD §7.1).
export function callEffectClass(call: CallSpec): EffectClass {
  if (call.effect) return call.effect;
  return isValidMethod(call.method) ? effectForMethod(call.method) : "unsafe";
}

/// One pipeline step. Exactly one of `call` / `transform` / `pipeline` /
/// `split` must be present (`validateSpec`).
export interface Step {
  condition: string | undefined;
  call: CallSpec | undefined;
  transform: Json | undefined;
  pipeline: PipelineSpec | undefined;
  split: Splitter | undefined;
  tryMode: boolean;
  /// Add the pipeline mount's `elevate` role to this call's principal.
  elevate: boolean;
  /// `as: "$name"` captures the step result into a variable.
  capture: string | undefined;
  /// Message name, keying `jsonObject` joins.
  name: string | undefined;
  retry: RetryPolicy | undefined;
}

export interface PipelineSpec {
  mode: Mode;
  onFail: Action | undefined;
  onSucceed: Action | undefined;
  concurrency: number | undefined;
  join: Joiner | undefined;
  steps: Step[];
}

export function emptyStep(): Step {
  return {
    condition: undefined,
    call: undefined,
    transform: undefined,
    pipeline: undefined,
    split: undefined,
    tryMode: false,
    elevate: false,
    capture: undefined,
    name: undefined,
    retry: undefined,
  };
}

export function emptySpec(): PipelineSpec {
  return { mode: "serial", onFail: undefined, onSucceed: undefined, concurrency: undefined, join: undefined, steps: [] };
}

export function callStep(method: string, url: string): Step {
  return { ...emptyStep(), call: { method, url, effect: undefined, headers: undefined } };
}

/// The effect class this step contributes to segment analysis.
export function stepEffectClass(step: Step): EffectClass | undefined {
  return step.call ? callEffectClass(step.call) : undefined;
}

/// Effective on-fail action: configured, else `stop` for parallel, `next`
/// otherwise (Restspace semantics retained).
export function failAction(spec: PipelineSpec): Action {
  return spec.onFail ?? (spec.mode === "parallel" ? "stop" : "next");
}

export function succeedAction(spec: PipelineSpec): Action {
  return spec.onSucceed ?? (spec.mode === "conditional" ? "end" : "next");
}

/// `http::Method::from_bytes`: any non-empty HTTP token.
export function isValidMethod(method: string): boolean {
  return /^[!#$%&'*+\-.^_`|~0-9A-Za-z]+$/.test(method);
}

/// `HeaderName::try_from`: a non-empty token (case-insensitive).
export function isValidHeaderName(name: string): boolean {
  return /^[!#$%&'*+\-.^_`|~0-9A-Za-z]+$/.test(name);
}

// ---- parsing (serde semantics) -----------------------------------------

class SpecParseError extends Error {}

function isObject(v: Json | undefined): v is JsonObject {
  return v !== undefined && v !== null && typeof v === "object" && !Array.isArray(v);
}

function enumField<T extends string>(obj: JsonObject, key: string, allowed: T[]): T | undefined {
  const v = obj[key];
  if (v === undefined || v === null) return undefined;
  if (typeof v !== "string" || !allowed.includes(v as T)) {
    throw new SpecParseError(`unknown variant ${JSON.stringify(v)} for '${key}', expected one of ${allowed.join(", ")}`);
  }
  return v as T;
}

function boolField(obj: JsonObject, key: string): boolean {
  const v = obj[key];
  if (v === undefined || v === null) return false;
  if (typeof v !== "boolean") throw new SpecParseError(`invalid type for '${key}': expected a boolean`);
  return v;
}

function stringField(obj: JsonObject, key: string): string | undefined {
  const v = obj[key];
  if (v === undefined || v === null) return undefined;
  if (typeof v !== "string") throw new SpecParseError(`invalid type for '${key}': expected a string`);
  return v;
}

function parseCall(v: Json): CallSpec {
  if (!isObject(v)) throw new SpecParseError("invalid type for 'call': expected an object");
  const method = v.method;
  const url = v.url;
  if (typeof method !== "string") throw new SpecParseError("missing field 'method' in call");
  if (typeof url !== "string") throw new SpecParseError("missing field 'url' in call");
  let headers: Record<string, string> | undefined;
  if (v.headers !== undefined && v.headers !== null) {
    if (!isObject(v.headers)) throw new SpecParseError("invalid type for 'headers': expected an object");
    headers = {};
    for (const [k, hv] of Object.entries(v.headers)) {
      if (typeof hv !== "string") throw new SpecParseError(`invalid type for header '${k}': expected a string`);
      headers[k] = hv;
    }
  }
  return { method, url, effect: enumField(v, "effect", EFFECTS), headers };
}

function parseStep(v: Json): Step {
  if (!isObject(v)) throw new SpecParseError("invalid type for step: expected an object");
  const step = emptyStep();
  step.condition = stringField(v, "if");
  if (v.call !== undefined && v.call !== null) step.call = parseCall(v.call);
  if (v.transform !== undefined) step.transform = v.transform;
  if (v.pipeline !== undefined && v.pipeline !== null) step.pipeline = parseSpecValue(v.pipeline);
  step.split = enumField(v, "split", ["jsonSplit"]);
  step.tryMode = boolField(v, "try");
  step.elevate = boolField(v, "elevate");
  step.capture = stringField(v, "as");
  step.name = stringField(v, "name");
  if (v.retry !== undefined && v.retry !== null) {
    const retry = retryFromConfig(v.retry);
    if (!retry) throw new SpecParseError("invalid 'retry' policy");
    step.retry = retry;
  }
  return step;
}

function parseSpecValue(v: Json): PipelineSpec {
  if (!isObject(v)) throw new SpecParseError("invalid type: expected a pipeline object");
  const spec = emptySpec();
  spec.mode = enumField(v, "mode", MODES) ?? "serial";
  spec.onFail = enumField(v, "onFail", ACTIONS);
  spec.onSucceed = enumField(v, "onSucceed", ACTIONS);
  if (v.concurrency !== undefined && v.concurrency !== null) {
    const c = v.concurrency;
    if (typeof c !== "number" || !Number.isInteger(c) || c < 0) {
      throw new SpecParseError("invalid type for 'concurrency': expected a non-negative integer");
    }
    spec.concurrency = c;
  }
  spec.join = enumField(v, "join", ["jsonObject"]);
  if (v.steps !== undefined && v.steps !== null) {
    if (!Array.isArray(v.steps)) throw new SpecParseError("invalid type for 'steps': expected an array");
    spec.steps = v.steps.map(parseStep);
  }
  return spec;
}

/// Deserialize the typed form (serde `from_value`), without validation.
export function specFromJson(value: Json): PipelineSpec {
  try {
    return parseSpecValue(value);
  } catch (e) {
    if (e instanceof SpecParseError) throw RsError.badRequest(`invalid pipeline spec: ${e.message}`);
    throw e;
  }
}

/// Parse from JSON: the typed form directly, or an array (the Restspace
/// string DSL, PRD §8.1) converted via `dsl.convert`. Validates.
export function specFromValue(value: Json, convertDsl: (v: Json) => PipelineSpec): PipelineSpec {
  const spec = Array.isArray(value) ? convertDsl(value) : specFromJson(value);
  const errors = validateSpec(spec);
  if (errors.length > 0) {
    throw RsError.validationFailed("pipeline spec failed validation", errors);
  }
  return spec;
}

// ---- serialization (serde `Serialize`, field order + skips) -------------

export function retryToJson(p: RetryPolicy): JsonObject {
  return {
    enabled: p.enabled,
    maxAttempts: p.maxAttempts,
    baseDelayMs: p.baseDelayMs,
    maxDelayMs: p.maxDelayMs,
    backoffMultiplier: p.backoffMultiplier,
    jitter: p.jitter,
    retryStatuses: [...p.retryStatuses],
    retryOnNetworkError: p.retryOnNetworkError,
    respectRetryAfter: p.respectRetryAfter,
  };
}

function callToJson(c: CallSpec): JsonObject {
  const out: JsonObject = { method: c.method, url: c.url };
  if (c.effect !== undefined) out.effect = c.effect;
  if (c.headers !== undefined) out.headers = { ...c.headers };
  return out;
}

function stepToJson(s: Step): JsonObject {
  const out: JsonObject = {};
  if (s.condition !== undefined) out.if = s.condition;
  if (s.call !== undefined) out.call = callToJson(s.call);
  if (s.transform !== undefined) out.transform = s.transform;
  if (s.pipeline !== undefined) out.pipeline = specToJson(s.pipeline);
  if (s.split !== undefined) out.split = s.split;
  if (s.tryMode) out.try = true;
  if (s.elevate) out.elevate = true;
  if (s.capture !== undefined) out.as = s.capture;
  if (s.name !== undefined) out.name = s.name;
  if (s.retry !== undefined) out.retry = retryToJson(s.retry);
  return out;
}

export function specToJson(spec: PipelineSpec): JsonObject {
  const out: JsonObject = { mode: spec.mode };
  if (spec.onFail !== undefined) out.onFail = spec.onFail;
  if (spec.onSucceed !== undefined) out.onSucceed = spec.onSucceed;
  if (spec.concurrency !== undefined) out.concurrency = spec.concurrency;
  if (spec.join !== undefined) out.join = spec.join;
  out.steps = spec.steps.map(stepToJson);
  return out;
}

// ---- validation ---------------------------------------------------------

/// Config-time validation: step shape, condition grammar, mode rules,
/// nested pipelines. Returns all problems, not just the first.
export function validateSpec(spec: PipelineSpec): string[] {
  const errors: string[] = [];
  validateInto(spec, "", errors);
  return errors;
}

function validateInto(spec: PipelineSpec, at: string, errors: string[]): void {
  if (spec.steps.length === 0) errors.push(`${at}: pipeline has no steps`);
  if (spec.mode === "parallel" && spec.onFail !== undefined && spec.onFail !== "stop") {
    errors.push(`${at}: parallel mode must stop on fail`);
  }
  if (spec.onSucceed === "stop") {
    errors.push(`${at}: cannot stop on success or the pipeline always fails`);
  }
  spec.steps.forEach((step, i) => {
    const here = `${at}/steps[${i}]`;
    const actions = [step.call, step.transform, step.pipeline, step.split].filter((p) => p !== undefined).length;
    if (actions !== 1) errors.push(`${here}: a step must have exactly one of call/transform/pipeline/split`);
    if (step.condition !== undefined) {
      try {
        parseCondition(step.condition);
      } catch (e) {
        errors.push(`${here}: invalid condition '${step.condition}': ${(e as Error).message}`);
      }
    }
    if (step.capture !== undefined && !step.capture.startsWith("$")) {
      errors.push(`${here}: 'as' must name a variable starting with '$'`);
    }
    if (step.call) {
      if (!isValidMethod(step.call.method)) errors.push(`${here}: invalid method '${step.call.method}'`);
      try {
        validatePattern(step.call.url);
      } catch (e) {
        errors.push(`${here}: ${(e as RsError).detail}`);
      }
      if (step.call.headers) {
        for (const [name, value] of Object.entries(step.call.headers)) {
          if (!isValidHeaderName(name)) errors.push(`${here}: invalid header name '${name}'`);
          try {
            validatePattern(value);
          } catch (e) {
            errors.push(`${here}: in header '${name}': ${(e as RsError).detail}`);
          }
        }
      }
    }
    if (step.transform !== undefined) validateTemplate(step.transform, here, errors);
    if (step.elevate && !step.call) errors.push(`${here}: 'elevate' applies only to call steps`);
    if (step.pipeline) validateInto(step.pipeline, here, errors);
  });
}

/// Walk a transform template with the same shape semantics as the runtime
/// apply (string leaves are JSONata expressions; objects and arrays recurse)
/// and record every leaf that fails to parse.
function validateTemplate(template: Json, at: string, errors: string[]): void {
  if (typeof template === "string") {
    try {
      validateExpr(template);
    } catch (e) {
      errors.push(`${at}: ${(e as RsError).detail}`);
    }
  } else if (Array.isArray(template)) {
    for (const v of template) validateTemplate(v, at, errors);
  } else if (template && typeof template === "object") {
    for (const v of Object.values(template)) validateTemplate(v, at, errors);
  }
}

export function countSteps(spec: PipelineSpec): number {
  return spec.steps.reduce((n, s) => n + 1 + (s.pipeline ? countSteps(s.pipeline) : 0), 0);
}
